### PR TITLE
Add `REDUNDANT_IMPORTS` lint for new redundant import detection

### PR DESCRIPTION
**Pull request by compiler-errors**
Defaults to Allow for now. Stacked on #123744 to avoid merge conflict, but much easier to review all as one.

r? petrochenkov

### CODE REVIEW

**Comment by compiler-errors** [comment]
nominating for T-lang to see if we'd like to keep this as allow or make it warn, etc., and perhaps to follow-up on the policy.

for the record, I think this lint is perhaps noisy but also *very* useful, so I'd like to see it at Warn level eventually. just not curious what the best rollout strategy is for that.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -419,6 +420,29 @@ declare_lint! {
     "imports that are never used"
 }
 
+declare_lint! {
+    /// The `unused_imports` lint detects imports that are redundant due to being
```
Comment (kadiwa4) [inline compiler/rustc_lint_defs/src/builtin.rs:424]: *`redundant_imports`

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in the lang meeting today.  We came to the consensus that we want to split out into a separate `allow-by-default` lint (or, alternatively, not lint at all) the case of an import that is redundant due to a glob import.

There are two reasons for this.

One, and most importantly, is that if we were to not do this, a `warn-by-default` lint here would make adding functions, statics, or consts to a prelude (including the one in the standard library) so disruptive as to have to be treated almost like a breaking change.

Two is that not doing this would make it annoying to migrate *away* from a glob import incrementally.

Once this case is separated out, we're OK with the remainder of this lint going directly to `warn-by-default`.

**Comment by petrochenkov** [comment]
Let's first land the renaming and the warn -> allow change, and then change the lint logic itself.
@rustbot author

**Comment by compiler-errors** [comment]
OK, I fixed the one nit. Haven't split/suppressed the glob behavior. I can open an issue to do that later.

@rustbot ready

**Comment by petrochenkov** [comment]
r=me with https://github.com/rust-lang/rust/pull/123813#discussion_r1582070070 addressed.
@rustbot author

**Comment by compiler-errors** [comment]
What the heck, I thought I literally fixed that. Sorry haha

**Pull request merged**